### DISCUSSION TITLE
Is C++ Windows Runtime Component (WinUI3) usable in WinUI3 Desktop Packaged project?

### DISCUSSION THREAD

**Discussion by hereafter** [microsoft/WindowsAppSDK#763] [Q&A] [answered]
I started using Project Reunion 0.5.5, samples compiled successfully.

Here is my setup
1. First Create an App1 project with "Blank App, Packaged (WinUI3 in Desktop)", it builds and runs successfully.
2. Create a Controls1 project with  "Windows Runtime Component (WinUI3"), both projects now build successfully.

3. Add a UserControl "UserControl" with the "User Control (WinUI3)" template, both projects now build successfully.

4. In "App1" Project and reference to the "Control1" project, now the build complains the following errors

Severity	Code	Description	Project	File	Line	Suppression State
Error	C2039	'Controls1': is not a member of 'winrt'	App1	C:\Users\hereafter\Desktop\App1\App1\App1\Generated Files\XamlTypeInfo.g.cpp	317	
Error	C3083	'Controls1': the symbol to the left of a '::' must be a type	App1	C:\Users\hereafter\Desktop\App1\App1\App1\Generated Files\XamlTypeInfo.g.cpp	317	
Error	C2039	'XamlMetaDataProvider': is not a member of 'winrt'	App1	C:\Users\hereafter\Desktop\App1\App1\App1\Generated Files\XamlTypeInfo.g.cpp	317	
Error	C2259	'winrt::App1::implementation::XamlMetaDataProvider': cannot instantiate abstract class	App1	C:\Users\hereafter\Desktop\App1\App1\App1\Generated Files\XamlTypeInfo.g.cpp	317	

What is going on here?

![image](https://user-images.githubusercontent.com/1483954/116029240-65972000-a68b-11eb-9a14-de68c66cbe52.png)

**Comment by andrewleader** [comment]
@DefaultRyan can you try out this repro? Notice that this involves creating two projects (a separate Runtime Component project).

**Reply by DefaultRyan** [reply]
Looking at it...

**Comment by DefaultRyan** [comment, marked as answer]
The Xaml Compiler knows enough about the project reference to add a reference to the runtime component's XamlMetaDataProvider:

```c++
    if (_otherProviders.empty())
    {
        _otherProviders.push_back(::winrt::Controls1::XamlMetaDataProvider());
        _otherProviders.push_back(::winrt::Microsoft::UI::Xaml::XamlTypeInfo::XamlControlsXamlMetaDataProvider());
    }
```

Unfortunately, it didn't add the needed #include statement to that file. I've filed an issue with WinUI to track this:
- [XamlTypeInfo.g.cpp adds providers for project references, but doesn't #include needed headers](https://github.com/microsoft/microsoft-ui-xaml/issues/4907)

XamlTypeInfo.g.cpp adds providers for project references, but doesn't #include needed headers
In the meantime, you can easily work around this by adding the needed #include to your precompiled header (pch.h):
```c++
#include <winrt/Controls1.h>
```

**Reply by hereafter** [reply]
Wow! 

tried to add to header, but everytime it got replaced by the generator.
pch.h never came to mind

Thanks
